src/Veldrid/MTL/MTLGraphicsDevice.cs

```diff
@@ -58,6 +58,25 @@ public MTLGraphicsDevice(GraphicsDeviceOptions options, SwapchainDescription? sw
         {
         }
 
+        public override void UpdateActiveDisplay(int x, int y, int w, int h)
+        {
+            if (_displayLink != null)
+            {
+                _frameEndedEvent.Set();
```

Comment:
This is definitely not the way to go about this. This could *potentially* cause permanent desynchronisation between callback and V-sync, at least according to my knowledge on how `CADisplayLink` works (not sure how that goes for `CVDisplayLink`). And is generally a hacky bandaid at best.

This suggests that the framework-side code isn't calling `UpdateActiveDisplay` at a good time, it must be called in the `DrawThread` in a safe place where the display link isn't blocked.

Here's an `osu-framework` diff that does this a bit smarter:
```diff
diff --git a/osu.Framework/Graphics/Veldrid/VeldridRenderer.cs b/osu.Framework/Graphics/Veldrid/VeldridRenderer.cs
index cc7d06722..6eedb6bf3 100644
--- a/osu.Framework/Graphics/Veldrid/VeldridRenderer.cs
+++ b/osu.Framework/Graphics/Veldrid/VeldridRenderer.cs
@@ -199,8 +199,11 @@ protected override void Initialise(IGraphicsSurface graphicsSurface)
             BufferUpdateCommands = Factory.CreateCommandList();
 
             pipeline.Outputs = Device.SwapchainFramebuffer.OutputDescription;
+
+            graphicsSurface.ActiveDisplayChanged += d => pendingActiveDisplay = d;
         }
 
+        private Display? pendingActiveDisplay;
         private Vector2 currentSize;
 
         protected internal override void BeginFrame(Vector2 windowSize)
@@ -232,7 +235,17 @@ protected internal override void FinishFrame()
             Device.SubmitCommands(Commands);
         }
 
-        protected internal override void SwapBuffers() => Device.SwapBuffers();
+        protected internal override void SwapBuffers()
+        {
+            Device.SwapBuffers();
+
+            if (pendingActiveDisplay != null)
+            {
+                Device.UpdateActiveDisplay(pendingActiveDisplay.Bounds.X, pendingActiveDisplay.Bounds.Y, pendingActiveDisplay.Bounds.Width, pendingActiveDisplay.Bounds.Height);
+                pendingActiveDisplay = null;
+            }
+        }
+
         protected internal override void WaitUntilIdle() => Device.WaitForIdle();
         protected internal override void WaitUntilNextFrameReady() => Device.WaitForNextFrameReady();
 
diff --git a/osu.Framework/Platform/IGraphicsSurface.cs b/osu.Framework/Platform/IGraphicsSurface.cs
index cc4c31562..ab29ec095 100644
--- a/osu.Framework/Platform/IGraphicsSurface.cs
+++ b/osu.Framework/Platform/IGraphicsSurface.cs
@@ -11,6 +11,8 @@ namespace osu.Framework.Platform
     /// </summary>
     public interface IGraphicsSurface
     {
+        event Action<Display>? ActiveDisplayChanged;
+
         /// <summary>
         /// A pointer representing a handle to this window, provided by the operating system.
         /// </summary>
diff --git a/osu.Framework/Platform/OsuTKGraphicsSurface.cs b/osu.Framework/Platform/OsuTKGraphicsSurface.cs
index 11367fa4e..77999fc08 100644
--- a/osu.Framework/Platform/OsuTKGraphicsSurface.cs
+++ b/osu.Framework/Platform/OsuTKGraphicsSurface.cs
@@ -17,6 +17,7 @@ public class OsuTKGraphicsSurface : IGraphicsSurface, IOpenGLGraphicsSurface
 
         public GraphicsSurfaceType Type => GraphicsSurfaceType.OpenGL;
 
+        public event Action<Display> ActiveDisplayChanged;
         public IntPtr WindowHandle => window.WindowInfo.Handle;
         public IntPtr DisplayHandle => throw new NotSupportedException($@"{nameof(DisplayHandle)} is not supported.");
 
diff --git a/osu.Framework/Platform/SDL2/SDL2GraphicsSurface.cs b/osu.Framework/Platform/SDL2/SDL2GraphicsSurface.cs
index 0c13b56ef..e66b410c1 100644
--- a/osu.Framework/Platform/SDL2/SDL2GraphicsSurface.cs
+++ b/osu.Framework/Platform/SDL2/SDL2GraphicsSurface.cs
@@ -19,6 +19,8 @@ public class SDL2GraphicsSurface : IGraphicsSurface, IOpenGLGraphicsSurface, IMe
 
         private IntPtr context;
 
+        public event Action<Display>? ActiveDisplayChanged;
+
         public IntPtr WindowHandle => window.WindowHandle;
         public IntPtr DisplayHandle => window.DisplayHandle;
 
@@ -84,6 +86,11 @@ public Size GetDrawableSize()
             return new Size(width, height);
         }
 
+        public void UpdateActiveDisplay(Display display)
+        {
+            ActiveDisplayChanged?.Invoke(display);
+        }
+
         #region OpenGL-specific implementation
 
         private void initialiseOpenGL()
diff --git a/osu.Framework/Platform/SDL2Window_Windowing.cs b/osu.Framework/Platform/SDL2Window_Windowing.cs
index debf407df..1adb46472 100644
--- a/osu.Framework/Platform/SDL2Window_Windowing.cs
+++ b/osu.Framework/Platform/SDL2Window_Windowing.cs
@@ -27,6 +27,7 @@ private void setupWindowing(FrameworkConfigManager config)
             CurrentDisplayBindable.Default = PrimaryDisplay;
             CurrentDisplayBindable.ValueChanged += evt =>
             {
+                graphicsSurface.UpdateActiveDisplay(evt.NewValue);
                 windowDisplayIndexBindable.Value = (DisplayIndex)evt.NewValue.Index;
             };
 
```

cc @peppy

Reply:
Is there a reason for putting it inside `SwapBuffers`? I'd probably put it in `BeginFrame`. That also goes for `AllowTearing` which is currently sitting arbitrarily inside `GameHost.DrawFrame()`

Reply:
That's what I had at first, but unfortunately `WaitForNextFrameReady` is called before `BeginFrame`, so we'll have to move the call to happen inside `BeginFrame`, which doesn't sound optimal.

Reply:
Oh... Probably want a separate method into `Renderer` from the exact point `WaitForNextFrameReady` is (i.e. it _must_ be before querying/spinning on the triple buffer). I still don't feel like `SwapBuffer` is the correct place.

Reply:
I don't understand @frenzibyte's concerns with the original logic.

That said, I think this whole approach is fundamentally flawed.

Logically: if a frame hasn't been *started*, the callback should not wait for it to *end*. So I've fixed this by making the semaphore manual reset, and added a reset call in `WaitForNextFrameReady`. I've tested this to work with the original proposed framework diff, and I've also cross-tested with https://github.com/ppy/osu-framework/pull/5846

Reply:
> I don't understand @frenzibyte's concerns with the original logic.

Right, I checked back on what I said and it turns out I misunderstood the logic a bit. With the code above, the callback can only exit early if `UpdateActiveDisplay` was called in the middle of a draw frame, so it won't have any major effect on synchronisation. The case I was thinking of was if the callback waits longer than the next V-sync, at which desynchronisation happens I guess ([read attached image](https://discord.com/channels/188630481301012481/589331078574112768/1119684304096743424)).